### PR TITLE
Secure vs insecure image pruning

### LINKED ISSUE #4232: pruning images - note the need for `--certificate-authority` flag when the registry is secured
Emphasize that the `oadm prune images` need to be passed `--certificate-authority` flag if the registry is secured and the current user's authority doesn't match the one used for the secured registry.

Don't forget to mention, what error can be seen if this flag is omitted. List it among common pruning issues.

### PR DESCRIPTION
**Pull request by miminar**
Document new options related to secure connection to integrated docker registry and a mechanism that decides whether to fall-back to insecure connection.

Resolves #4232
Resolves [bz#1469654](https://bugzilla.redhat.com/show_bug.cgi?id=1469654)

~~Is blocked on openshift/origin#14114?~~ No longer blocked.

### CODE REVIEW

**Comment by miminar** [comment]
@legionus PTAL

**Comment by miminar** [comment]
@soltysh PTAL

File: admin_guide/pruning_resources.adoc [hunk 1]
```diff
@@ -283,9 +306,39 @@ $ oadm prune images --keep-tag-revisions=3 --keep-younger-than=60m --confirm
...
+
+To always force secure connection, specify `--certificate-authority`.
+
+[IMPORTANT]
+====
+If the registry is secured by a certificate authority different from the one
+used by {product-title}, it needs to be specified using
+`--certificate-authority` flag. Otherwise, the prune command will fail with
+error similar to
+link:#using-insecure-connection-against-secured-registry[the one below].
```
Comment (ahardin-rh) [inline admin_guide/pruning_resources.adoc]: xref:using-insecure-connection-against-secured-registry[the one below]

File: admin_guide/pruning_resources.adoc [hunk 2]
```diff
@@ -188,14 +188,27 @@ particular namespace makes it impossible to calculate their current usage.
...
+from the current user's config file. If provided, secure connection will be
+initiated.
 
 .^|`--confirm`
 |Indicate that pruning should occur, instead of performing a dry-run.
 
+.^|`--force-insecure`
+|Allow an insecure connection to the docker registry that is hosted
+via HTTP or has an invalid HTTPS certificate. By default, insecure connection
+is allowed only in case where neither certificate-authority nor registry-url is
```
Comment (ahardin-rh) [inline admin_guide/pruning_resources.adoc]: `certificate-authority`
    `registry-url`

File: admin_guide/pruning_resources.adoc [hunk 3]
```diff
@@ -283,9 +306,39 @@ $ oadm prune images --keep-tag-revisions=3 --keep-younger-than=60m --confirm
...
+insecure connection when prompted.
+3. The registry url is not provided or it is a local-link address or localhost.
+
+To always force secure connection, specify `--certificate-authority`.
+
+[IMPORTANT]
+====
+If the registry is secured by a certificate authority different from the one
+used by {product-title}, it needs to be specified using
+`--certificate-authority` flag. Otherwise, the prune command will fail with
```
Comment (ahardin-rh) [inline admin_guide/pruning_resources.adoc]: `prune` command

s/with error similar to/with an error similar to

File: admin_guide/pruning_resources.adoc [hunk 4]
```diff
@@ -337,3 +390,37 @@ Such tags tend to have just one image in its history, which effectively prevents
...
+
+----
+error: error communicating with registry: Get http://172.30.30.30:5000/healthz: malformed HTTP response "\x15\x03\x01\x00\x02\x02"
+error: error communicating with registry: [Get https://172.30.30.30:5000/healthz: x509: certificate signed by unknown authority, Get http://172.30.30.30:5000/healthz: malformed HTTP response "\x15\x03\x01\x00\x02\x02"]
+----
+
+It means that your registry is secured using a certificate signed by a
+certificate authority other than the one used by `oadm prune images` client for
+connection verification. By default, the certificate authority data stored in
+user's config are used - the same for communication with the master API.
```
Comment (ahardin-rh) [inline admin_guide/pruning_resources.adoc]: s/config/configuration file
s/-/--

File: admin_guide/pruning_resources.adoc [hunk 5]
```diff
@@ -337,3 +390,37 @@ Such tags tend to have just one image in its history, which effectively prevents
...
+command:
+
+----
+error: error communicating with registry: Get http://172.30.30.30:5000/healthz: malformed HTTP response "\x15\x03\x01\x00\x02\x02"
+error: error communicating with registry: [Get https://172.30.30.30:5000/healthz: x509: certificate signed by unknown authority, Get http://172.30.30.30:5000/healthz: malformed HTTP response "\x15\x03\x01\x00\x02\x02"]
+----
+
+It means that your registry is secured using a certificate signed by a
+certificate authority other than the one used by `oadm prune images` client for
+connection verification. By default, the certificate authority data stored in
```
Comment (ahardin-rh) [inline admin_guide/pruning_resources.adoc]: I would move the code block to just before "By default...." so that the first sentence does not get broken up.

File: admin_guide/pruning_resources.adoc [hunk 6]
```diff
@@ -188,14 +188,27 @@ particular namespace makes it impossible to calculate their current usage.
 
 |Option |Description
 
+.^|`--all`
+|Include images that were not pushed to the registry but have been mirrored by
```
Comment (ahardin-rh) [inline admin_guide/pruning_resources.adoc]: registry, but

File: admin_guide/pruning_resources.adoc [hunk 7]
```diff
@@ -283,9 +306,39 @@ $ oadm prune images --keep-tag-revisions=3 --keep-younger-than=60m --confirm
 $ oadm prune images --prune-over-size-limit --confirm
 ----
 
+.Using secure or insecure connection
```
Comment (ahardin-rh) [inline admin_guide/pruning_resources.adoc]: Perhaps we can use [discrete headings ](https://github.com/openshift/openshift-docs/blob/master/contributing_to_docs/doc_guidelines.adoc#discrete-headings)here instead?

File: admin_guide/pruning_resources.adoc [hunk 8]
```diff
@@ -209,8 +222,18 @@ is younger than `<duration>` relative to the current time. (default *60m*)
...
+.^|`--registry-url`
+|The address to use when contacting the registry. The command will attempt to
+use a cluster-internal URL determined from managed images and image streams. In
+case it fails (the registry cannot be resolved or reached), an alternative route
+that works needs to be provided using this flag. If provided, a secure
+connection will be initiated. To allow a fall-back to insecure transport, use
+`--force-insecure`.
 |===
 
+.Determining images and layers eligible for pruning
```
Comment (ahardin-rh) [inline admin_guide/pruning_resources.adoc]: Perhaps we can use a [discrete heading](https://github.com/openshift/openshift-docs/blob/master/contributing_to_docs/doc_guidelines.adoc#discrete-headings)here instead?

File: admin_guide/pruning_resources.adoc [hunk 9]
```diff
@@ -188,14 +188,27 @@ particular namespace makes it impossible to calculate their current usage.
...
+initiated.
 
 .^|`--confirm`
 |Indicate that pruning should occur, instead of performing a dry-run.
 
+.^|`--force-insecure`
+|Allow an insecure connection to the docker registry that is hosted
+via HTTP or has an invalid HTTPS certificate. By default, insecure connection
+is allowed only in case where neither certificate-authority nor registry-url is
+specified. Whenever possible, use `--certificate-authority` instead of this
```
Comment (soltysh) [inline admin_guide/pruning_resources.adoc]: I'd move the last sentence as a NOTE below. Additionally, reword this to something like:

Whenever possible use `--certificate-authority`, instead. Use of this option is strongly discouraged.

File: admin_guide/pruning_resources.adoc [hunk 10]
```diff
@@ -209,8 +222,18 @@ is younger than `<duration>` relative to the current time. (default *60m*)
 |Prune each image that exceeds the smallest xref:limits.adoc#image-limits[limit]
 defined in the same project. This flag cannot be combined with `--keep-tag-revisions`
 nor `--keep-younger-than`.
+
+.^|`--registry-url`
+|The address to use when contacting the registry. The command will attempt to
+use a cluster-internal URL determined from managed images and image streams. In
+case it fails (the registry cannot be resolved or reached), an alternative route
+that works needs to be provided using this flag. If provided, a secure
+connection will be initiated. To allow a fall-back to insecure transport, use
```
Comment (soltysh) [inline admin_guide/pruning_resources.adoc]: Don't encourage using `--force-insecure` anywhere, so drop the last sentence. It's already explained and that's it, it should not be used anywhere, though.

File: admin_guide/pruning_resources.adoc [hunk 11]
```diff
@@ -283,9 +306,39 @@ $ oadm prune images --keep-tag-revisions=3 --keep-younger-than=60m --confirm
 $ oadm prune images --prune-over-size-limit --confirm
 ----
 
+.Using secure or insecure connection
+
+The secure connection is preferred for production use. It is done over HTTPS
```
Comment (soltysh) [inline admin_guide/pruning_resources.adoc]: Suggestion:
The secure connection is the preferred and recommended approach.

It's the recommended as a rule, not only for production.

File: admin_guide/pruning_resources.adoc [hunk 12]
```diff
@@ -283,9 +306,39 @@ $ oadm prune images --keep-tag-revisions=3 --keep-younger-than=60m --confirm
...
+protocol with a mandatory certificate verification. The prune command always
+attempts to use it if possible. If not possible, in some cases it can fall-back
+to insecure connection which is dangerous. In this case, either certificate
+verification is skipped or plain HTTP protocol is used.
+
+The fall-back to insecure connection is allowed in following cases unless
```
Comment (soltysh) [inline admin_guide/pruning_resources.adoc]: s/following/the following

Comment (soltysh) [inline admin_guide/pruning_resources.adoc]: following cases **,** unless

File: admin_guide/pruning_resources.adoc [hunk 13]
```diff
@@ -283,9 +306,39 @@ $ oadm prune images --keep-tag-revisions=3 --keep-younger-than=60m --confirm
...
+`--certicate-authority` is specified:
+
+1. The prune command is run with `--force-insecure` option.
```
Comment (soltysh) [inline admin_guide/pruning_resources.adoc]: Add info in parenthesis this is not recommended.

File: admin_guide/pruning_resources.adoc [hunk 14]
```diff
@@ -337,3 +390,37 @@ Such tags tend to have just one image in its history, which effectively prevents
...
+
+----
+error: error communicating with registry: Get https://172.30.30.30:5000/healthz: http: server gave HTTP response to HTTPS client
+----
+
+It means that your registry is not secured and the `oadm prune images` client
+attempts to use secure connection. The recommened solution is to
+xref:../install_config/registry/securing_and_exposing_registry.adoc#securing-the-registry[secure
+the registry]. If that is not desired, you may force the client to use insecure
+connection by appending `--force-insecure` to the command.
```
Comment (soltysh) [inline admin_guide/pruning_resources.adoc]: Add not recommended at the end in parens.

**Review by soltysh — CHANGES_REQUESTED** [review]
I left you some comments.

**Comment by miminar** [comment]
Now blocked on openshift/origin#14405. I will need to update error messages once it lands.

**Comment by miminar** [comment]
New blocked on openshift/origin#14914

File: admin_guide/pruning_resources.adoc (hunk 9 above)
Comment (miminar) [inline admin_guide/pruning_resources.adoc]: I've ditched it. There is plenty of other places advocating the same. Instead, I've put *(Dangerous)* at the beginning.

**Comment by miminar** [comment]
Thanks for all the comments. They should be addressed now. The dependency PR is still waiting for the unblocked merge queue.

**Comment by miminar** [comment]
No longer blocked. @ahardin-rh could you please review once more?

**Comment by bparees** [comment]
@miminar if we label this PR for 3.5 and 3.6, the changes will be applied there.
 
if 3.5 and 3.6 need a *different* set of doc, then you'll need to open a separate PR and only label it for 3.5 and 3.6.

**Comment by bparees** [comment]
@miminar so which versions is this specific PR appropriate for?

**Comment by miminar** [comment]
@bparees 3.6 - latest; I'll re-submit new PRs for 3.4 and 3.5

**Comment by bparees** [comment]
@miminar ok.

@openshift/team-documentation ptal/merge.

**Comment by miminar** [comment]
I did some small corrections to the descriptions of flags.

File: admin_guide/pruning_resources.adoc [hunk 15]
```diff
@@ -191,13 +191,27 @@ particular namespace makes it impossible to calculate their current usage.
...
 
 .^|`--confirm`
-|Indicate that pruning should occur, instead of performing a dry-run.
+|Indicate that pruning should occur, instead of performing a dry-run. This
+requires a valid route to the integrated Docker registry. If this command is
+run outside of the cluster network, the route needs to be provided using
+`--registry-url`.
+
+.^|`--force-insecure`
+|*(Dangerous)* Allow an insecure connection to the docker registry that is
```
Comment (ahardin-rh) [inline admin_guide/pruning_resources.adoc]: Instead of labeling this Dangerous, maybe we can say:

    *Use caution with this option.* Allow...

File: admin_guide/pruning_resources.adoc [hunk 16]
```diff
@@ -191,13 +191,27 @@ particular namespace makes it impossible to calculate their current usage.
...
+hosted via HTTP or has an invalid HTTPS certificate. See
+xref:#pruning-images-secure-or-insecure[below] for more information.
```
Comment (ahardin-rh) [inline admin_guide/pruning_resources.adoc]: s/below/Using Secure or Insecure Connections

File: admin_guide/pruning_resources.adoc [hunk 17]
```diff
@@ -287,9 +305,41 @@ $ oadm prune images --keep-tag-revisions=3 --keep-younger-than=60m --confirm
 $ oadm prune images --prune-over-size-limit --confirm
 ----
 
+[[pruning-images-secure-or-insecure]]
+=== Using secure or insecure connection
```
Comment (ahardin-rh) [inline admin_guide/pruning_resources.adoc]: Using Secure or Insecure Connections

File: admin_guide/pruning_resources.adoc [hunk 18]
```diff
@@ -287,9 +305,41 @@ $ oadm prune images --keep-tag-revisions=3 --keep-younger-than=60m --confirm
...
+
+The secure connection is preferred and recomended approach. It is done over
```
Comment (ahardin-rh) [inline admin_guide/pruning_resources.adoc]: is the preferred and recommended approach

File: admin_guide/pruning_resources.adoc [hunk 19]
```diff
@@ -287,9 +305,41 @@ $ oadm prune images --keep-tag-revisions=3 --keep-younger-than=60m --confirm
...
+HTTPS protocol with a mandatory certificate verification. The `prune` command
+always attempts to use it if possible. If not possible, in some cases it can
+fall-back to insecure connection which is dangerous. In this case, either
```
Comment (ahardin-rh) [inline admin_guide/pruning_resources.adoc]: connection, which is dangerous

File: admin_guide/pruning_resources.adoc [hunk 20]
```diff
@@ -287,9 +305,41 @@ $ oadm prune images --keep-tag-revisions=3 --keep-younger-than=60m --confirm
...
+certificate verification is skipped or plain HTTP protocol is used.
+
+The fall-back to insecure connection is allowed in the following cases unless
+`--certificate-authority` is specified:
+
+1. *(Dangerous)* The `prune` command is run with `--force-insecure` option.
```
Comment (ahardin-rh) [inline admin_guide/pruning_resources.adoc]: run with the

File: admin_guide/pruning_resources.adoc [hunk 21]
```diff
@@ -287,9 +305,41 @@ $ oadm prune images --keep-tag-revisions=3 --keep-younger-than=60m --confirm
...
+2. Provided registry-url is prefixed with `http://` scheme
```
Comment (ahardin-rh) [inline admin_guide/pruning_resources.adoc]: The provided `registry-url` is prefixed with the `http://` scheme.

File: admin_guide/pruning_resources.adoc [hunk 22]
```diff
@@ -287,9 +305,41 @@ $ oadm prune images --keep-tag-revisions=3 --keep-younger-than=60m --confirm
...
+3. Provided registry-url is a local-link address or localhost.
```
Comment (ahardin-rh) [inline admin_guide/pruning_resources.adoc]: The provided `registry-url`

File: admin_guide/pruning_resources.adoc [hunk 23]
```diff
@@ -287,9 +305,41 @@ $ oadm prune images --keep-tag-revisions=3 --keep-younger-than=60m --confirm
...
+4. Config of the current user allows for insecure connection. This may be
```
Comment (ahardin-rh) [inline admin_guide/pruning_resources.adoc]: The configuration of the current user allows for an insecure connection.

File: admin_guide/pruning_resources.adoc [hunk 24]
```diff
@@ -287,9 +305,41 @@ $ oadm prune images --keep-tag-revisions=3 --keep-younger-than=60m --confirm
...
+caused by the user either logging in using `--insecure-skip-tls-verify` or
+choosing the insecure connection when prompted.
+
+[IMPORTANT]
+====
+If the registry is secured by a certificate authority different from the one
+used by {product-title}, it needs to be specified using
+`--certificate-authority` flag. Otherwise, the `prune` command will fail with
```
Comment (ahardin-rh) [inline admin_guide/pruning_resources.adoc]: using the `--certificate-authority` flag.

File: admin_guide/pruning_resources.adoc [hunk 25]
```diff
@@ -287,9 +305,41 @@ $ oadm prune images --keep-tag-revisions=3 --keep-younger-than=60m --confirm
...
+an error similar to
+xref:using-wrong-certificate-authority[the one] or
```
Comment (ahardin-rh) [inline admin_guide/pruning_resources.adoc]: similar to those seen in xref:using-wrong-certificate-authority[Using the Wrong Certificate Authority] or xref:using-insecure-connection-against-secured-registry[Using an Insecure Connection Against a Secured Registry].

File: admin_guide/pruning_resources.adoc [hunk 26]
```diff
@@ -287,9 +305,41 @@ $ oadm prune images --keep-tag-revisions=3 --keep-younger-than=60m --confirm
...
+xref:using-insecure-connection-against-secured-registry[the other] below.
+====
+
 [[image-pruning-problems]]
 === Image Pruning Problems
 
+[discrete]
+==== Images not being pruned
```
Comment (ahardin-rh) [inline admin_guide/pruning_resources.adoc]: Images Not Being Pruned

File: admin_guide/pruning_resources.adoc [hunk 27]
```diff
@@ -342,6 +392,58 @@ them from ever being pruned.
 xref:../dev_guide/managing_images.adoc#tag-naming[Learn more about _istag_
 naming.]
 
+[discrete]
+[[using-secure-connection-against-insecure-registry]]
+==== Using secure connection against insecure registry
```
Comment (ahardin-rh) [inline admin_guide/pruning_resources.adoc]: Using a Secure Connection Against an Insecure Registry

File: admin_guide/pruning_resources.adoc [hunk 28]
```diff
@@ -342,6 +392,58 @@ them from ever being pruned.
...
+
+If you see similar message in the output of the `oadm prune images` command:
```
Comment (ahardin-rh) [inline admin_guide/pruning_resources.adoc]: If you see a message similar to the following in the output of the `oadm prune images` command, then your registry is not secured and the `oadm prune images` client will attempt to use a secure connection:

File: admin_guide/pruning_resources.adoc [hunk 29]
```diff
@@ -342,6 +392,58 @@ them from ever being pruned.
...
+
+----
+error: error communicating with registry: Get https://172.30.30.30:5000/healthz: http: server gave HTTP response to HTTPS client
+----
+
+It means that your registry is not secured and the `oadm prune images` client
+attempts to use secure connection. The recommened solution is to
+xref:../install_config/registry/securing_and_exposing_registry.adoc#securing-the-registry[secure
+the registry]. If that is not desired, you may force the client to use insecure
```
Comment (ahardin-rh) [inline admin_guide/pruning_resources.adoc]: you can force the client to use an insecure

File: admin_guide/pruning_resources.adoc [hunk 30]
```diff
@@ -342,6 +392,58 @@ them from ever being pruned.
...
+connection by appending `--force-insecure`  to the command *(not recommended)*.
+
+[[using-insecure-connection-against-secured-registry]]
+==== Using insecure connection against secured registry
```
Comment (ahardin-rh) [inline admin_guide/pruning_resources.adoc]: Using an Insecure Connection Against a Secured Registry

File: admin_guide/pruning_resources.adoc [hunk 31]
```diff
@@ -342,6 +392,58 @@ them from ever being pruned.
...
+
+----
+error: error communicating with registry: Get http://172.30.30.30:5000/healthz: malformed HTTP response "\x15\x03\x01\x00\x02\x02"
+error: error communicating with registry: [Get https://172.30.30.30:5000/healthz: x509: certificate signed by unknown authority, Get http://172.30.30.30:5000/healthz: malformed HTTP response "\x15\x03\x01\x00\x02\x02"]
+----
+
+By default, the certificate authority data stored in user's configuration file
+are used -- the same for communication with the master API.
+
+Use `--certificate-authority` option to provide the right certificate authority
```
Comment (ahardin-rh) [inline admin_guide/pruning_resources.adoc]: use the

File: admin_guide/pruning_resources.adoc [hunk 32]
```diff
@@ -342,6 +392,58 @@ them from ever being pruned.
...
+for the docker registry server.
+
+[discrete]
+[[using-wrong-certificate-authority]]
+==== Using wrong certificate authority
```
Comment (ahardin-rh) [inline admin_guide/pruning_resources.adoc]: Using the Wrong Certificate Authority

File: admin_guide/pruning_resources.adoc [hunk 33]
```diff
@@ -342,6 +392,58 @@ them from ever being pruned.
...
+
+The following error means that certificate authority used to sign the
```
Comment (ahardin-rh) [inline admin_guide/pruning_resources.adoc]: that the certificate

File: admin_guide/pruning_resources.adoc [hunk 34]
```diff
@@ -342,6 +392,58 @@ them from ever being pruned.
...
+certificate of the secured Docker registry is different from the authority used
```
Comment (ahardin-rh) [inline admin_guide/pruning_resources.adoc]: different than

File: admin_guide/pruning_resources.adoc [hunk 35]
```diff
@@ -342,6 +392,58 @@ them from ever being pruned.
...
+by the client.
+
+----
+error: error communicating with registry: Get https://172.30.30.30:5000/: x509: certificate signed by unknown authority
+----
+
+Make sure to provide the right one with the flag `--certificate-authority`.
+
+As a work-around, the `--force-insecure` flag may be added instead *(not
```
Comment (ahardin-rh) [inline admin_guide/pruning_resources.adoc]: can be added instead

File: admin_guide/pruning_resources.adoc (hunk 15 above)
Comment (ahardin-rh) [inline admin_guide/pruning_resources.adoc]: Docker registry

File: admin_guide/pruning_resources.adoc [hunk 36]
```diff
@@ -342,6 +392,58 @@ them from ever being pruned.
...
+error: error communicating with registry: Get http://172.30.30.30:5000/healthz: malformed HTTP response "\x15\x03\x01\x00\x02\x02"
+error: error communicating with registry: [Get https://172.30.30.30:5000/healthz: x509: certificate signed by unknown authority, Get http://172.30.30.30:5000/healthz: malformed HTTP response "\x15\x03\x01\x00\x02\x02"]
+----
+
+By default, the certificate authority data stored in user's configuration file
+are used -- the same for communication with the master API.
+
+Use `--certificate-authority` option to provide the right certificate authority
+for the docker registry server.
+
```
Comment (ahardin-rh) [inline admin_guide/pruning_resources.adoc]: Docker registry

File: admin_guide/pruning_resources.adoc (hunk 34 above)
Comment (ahardin-rh) [reply admin_guide/pruning_resources.adoc]: Docker registry

**Comment by ahardin-rh** [comment]
@miminar Thanks! Just a few minor comments from me. 

Just a heads-up that, given our new docs workflow, if you want to submit separate PRs for 3.4 and 3.5, be sure to do so against enterprise-3.4-stage and enterprise-3.5-stage respectively. Thanks again!

**Comment by miminar** [comment]
@ahardin-rh thanks a lot! Comments should be addressed now.

**Comment by ahardin-rh** [comment]
@miminar Excellent. Thank you!

**Comment by ahardin-rh** [comment]
[rev_history]
|xref:../admin_guide/pruning_resources.adoc#admin-guide-pruning-resources[Pruning Objects]
|Added details on secure versus insecure image pruning.
%
